docs/topics/guideline/solidity.md

```diff
@@ -0,0 +1,302 @@
+---
+description: Solidity best practices.
+---
+
+# General
+
+This document describes best practices and patterns to use when writing and porting solidity code.
+
+***This document was written in mid 2021 and current solidity version is 0.8.***
+
+The document is composed of three parts:
+
+1. The first one is a general recommendations for coding style
+2. The second part is a more opinionated set of design patterns and recommendations specific to solidity that helps with
+   the security of the code and are usually objected by auditors if not used.
+3. The third part are guidelines for contract upgrade from 0.5 -> 0.8
+
+# Coding Style
+
+This part is based heavily
+on: [Solidity Style Guide](https://docs.soliditylang.org/en/latest/style-guide.html#contract-and-library-names)
+
+In the code we heavily recommend using linter and prettier to force a common style.
+
+Below is a summary and some extra guidelines to follow:
+
+- [NatSpec](https://docs.soliditylang.org/en/latest/style-guide.html#natspec): **Every function and variable declaration
+  must have NatSpec comments**.
+- File structure:
+    - One contract, library or interface per file.
+    - [Order of Layout](https://docs.soliditylang.org/en/latest/style-guide.html#natspec): `Type declarations`,
+      `State variables`, `Events`, `Functions`.
+    - [Order of Functions](https://docs.soliditylang.org/en/latest/style-guide.html#order-of-functions):
+      `constructor`, `fallback function` (if exists), `external`, `public`, `internal`, `private`. Within a grouping,
+      place the view and pure functions last.
+- [Naming Styles](https://docs.soliditylang.org/en/latest/style-guide.html#naming-styles):
+    - Contracts, libraries, events, enums and structs should be named using the ***CapWords*** style and match their
+      filenames.
+    - Function, function arguments, local, state variable and modifiers should use ***mixedCase***. To avoid naming
+      collisions add a single trailing underscore `_` to function arguments. This is specially useful when assigning
+      function arguments to state variables.
+    - Constants should be named with all ***CAPITAL_LETTERS_WITH_UNDERSCORES*** separating words.
+    - For library functions that operate on a custom struct the first argument should always be named `self`.
+    - Directory names should use ***all_lowercase_with_underscores***. Favor breaking apart names into subdirectories,
```

Comment:
I'm in team hyphen for directories.
It's hard to find any examples with multiple directories but i found one on openzeppelin
https://docs.openzeppelin.com/learn/developing-smart-contracts#adding-more-contracts

```
// contracts/access-control/Auth.sol
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

contract Auth {
```
And it would be the same than nodejs